README.md
GRValidation
============

GRValidation is a validation toolkit for Swift 3.

It lets you validate both simple values and complex models, and won't let you down when your validations leave the trivial zone.

**March, 1 2017: GRValidation 0.3.0 is out** - [Release notes](CHANGELOG.md). Follow [@groue](http://twitter.com/groue) on Twitter for release announcements and usage tips.


Features
--------

- Type safety
- Value validation
- Value reparation
- Complex model validation
- Detailed error messages

Missing features so far:

- Localization of validation errors
- Introspection of validation errors


Value Validation vs. Model Validation
-------------------------------------

GRValidation distinguishes *Value Validation* from *Model Validation*.

Precisely speaking, **Value Validation** throws errors like "12 should be greater than 18", and is responsible for:

- **Value Checking**, as in "is this string empty?"
- **Value Reparation**: for example, let's trim, validate, and format this user-provided phone number.
- **Composition**: several validations can be composed in a single more complex one, as in "this string should be nil, or not empty".

**Model validation**, on the other side, throws errors like "age should be greater than 18". It is built on top of value validation and provides:

- **Property validation**, as in "name should not be empty."
- **Global validation**, as in "Please provide an email or a phone number."
- **Mutating validations**. This is value reparation, put to some use: for example, after successful validation of a user-provided phone number, one wants to store its formatted version.


Documentation
-------------

- [Value Validation](#value-validation)
    - [The ValidationType Protocol](#the-validationtype-protocol)
    - [ValidationError vs Boolean Check](#validationerror-vs-boolean-check)
    - [Built-in Value Validations](#built-in-value-validations)
    - [Composed Validations](#composed-validations)
- [Model Validation](#model-validation)
    - [The Validable Protocol](#the-validable-protocol)


## Value Validation

### The ValidationType Protocol

**ValidationType** is a protocol that checks a value of type TestedType, and eventually returns a value of type ValidType, or throws a ValidationError:

```swift
public protocol ValidationType {
    typealias TestedType
    typealias ValidType
    func validate(value: TestedType) throws -> ValidType
}
```

For example:

```swift
// Positive integer
let v = ValidationRange(minimum: 0)
try v.validate(1)   // OK: 1
try v.validate(nil) // ValidationError: nil should be greater than or equal to 0.
try v.validate(-1)  // ValidationError: -1 should be greater than or equal to 0.
```

The returned value may be different from the input:

```swift
enum Color : Int {
    case Red
    case White
    case Rose
}
let v = ValidationRawValue<Color>()
try v.validate(0)   // OK: Color.Red
try v.validate(3)   // ValidationError: 3 is an invalid Color.
```


### ValidationError vs Boolean Check

The validate() method may throw a ValidationError:

```swift
let positiveInt = ValidationRange(minimum: 0)
try positiveInt.validate(-1)    // Throws a ValidationError
```

You may also perform a simple boolean check with the `~=` operator, or via pattern matching:

```swift
positiveInt ~= 10  // true
positiveInt ~= -1  // false

switch int {
case positiveInt:
    // int passes validation
    ...
}
```


### Built-in Value Validations

| Validation type              | TestedType      | ValidType       |            |
|:---------------------------- |:--------------- |:--------------- |:---------- |
| Validation                   | T               | T               | All values pass. |
| ValidationFailure            | T               | T               | All values fail. |
| ValidationNil                | T?              | T?              | Checks that input is nil. |
| ValidationNotNil             | T?              | T               | Checks that input is not nil. |
| ValidationTrim               | String?         | String?         | All strings pass. Non nil strings are trimmed. |
| ValidationStringLength       | String?         | String          | Checks that a string is not nil and has length in a specific range. |
| ValidationRegularExpression  | String?         | String          | Checks that a string is not nil and matches a regular expression. |
| ValidationCollectionNotEmpty | CollectionType? | CollectionType  | Checks that a collection is not nil and not empty. |
| ValidationCollectionEmpty    | CollectionType? | CollectionType? | Checks that a collection is nil or empty. |
| ValidationEqual              | T? where T:Equatable | T          | Checks that a value is not nil and equal to a reference value. |
| ValidationNotEqual           | T? where T:Equatable | T?         | Checks that a value is nil or not equal to a reference value. |
| ValidationElementOf          | T? where T:Equatable | T          | Checks that a value is not nil and member of a reference collection. |
| ValidationNotElementOf       | T? where T:Equatable | T?         | Checks that a value is nil or not member of a reference collection. |
| ValidationRawValue           | T.RawValue? where T: RawRepresentable | T | Checks that a value is not nil and a valid raw value for type T. |
| ValidationRange              | T? where T: ForwardIndexType, T: Comparable | T | Checks that a value is not nil and in a specific range. |


### Composed Validations

Basic Value Validations can be chained, or composed using boolean operators:

- `v1 >>> v2`
    
    Chains two validations. Returns the value returned by v2.
    
    ```swift
    // Checks that a string matches a regular expression, after trimming:
    let v = ValidationTrim() >>> ValidationRegularExpression(pattern: "^[0-9]+$")
    try v.validate(" 123 ") // "123"
    try v.validate("foo")   // ValidationError
    ```
    
- `v1 || v2`
    
    Checks that a value passes at least one validation. Returns the value returned by the first validation that passes, or the input value when output types don't match.
    
    ```swift
    // Checks that an Int is not nil and equal to 1 or 2:
    let v = ValidationEqual(1) || ValidationEqual(2)
    try v.validate(1) // 1
    try v.validate(2) // 2
    try v.validate(3) // ValidationError
    ```
    
- `v1 && v2`
    
    Checks that a value passes both validations. Returns the value returned by v2.
    
    ```swift
    // Checks that an Int is nil, or not 1, and not 2:
    let v = ValidationNotEqual(1) && ValidationNotEqual(2)
    try v.validate(1) // ValidationError
    try v.validate(2) // ValidationError
    try v.validate(3) // 3
    ```
    
- `!v1`

    Inverts a validation. Returns the input value, or throws a generic "is invalid." error.
    
    ```swift
    // Checks that an Int is not 1.
    let v = !ValidationEqual(1)
    try v.validate(1) // ValidationError
    try v.validate(2) // 2
    ```


## Model Validation

### The Validable Protocol

The **Validable** protocol provides methods that help validating models.

Let's start with a simple model:

```swift
struct Person: Validable {
    var name: String?
    
    func validate() throws {
        // Name should not be nil or empty.
        try validate(property: "name", with: name >>> ValidationStringLength(minimum: 1))
    }
}

let person = Person(name: "Arthur")
try person.validate()   // OK

let person = Person(name: nil)
try person.validate()
// Invalid Person(name: nil): name should not be empty.
```


Feel at your ease, and don't hesitate building more complex validations:

```swift
struct Person : Validable {
    var name: String?
    var age: Int?
    var email: String?
    var phoneNumber: String?
    
    mutating func validate() throws {
        // ValidationPlan doesn't fail on the first validation error. Instead,
        // it gathers all of them, and eventually throws a single ValidationError.
        try ValidationPlan()
            .append {
                // Name should not be empty after whitespace trimming:
                let nameValidation = ValidationTrim() >>> ValidationStringLength(minimum: 1)
                name = try validate(
                    property: "name",
                    with: name >>> nameValidation)
            }
            .append {
                // Age should be nil, or positive:
                let ageValidation = ValidationNil() || ValidationRange(minimum: 0)
                try validate(
                    property: "age",
                    with: age >>> ageValidation)
            }
            .append {
                // Email should be nil, or contain @ after whitespace trimming:
                let emailValidation = ValidationNil() || (ValidationTrim() >>> ValidationRegularExpression(pattern:"@"))
                email = try validate(
                    property: "email",
                    with: email >>> emailValidation)
            }
            .append {
                // Phone number should be nil, or be a valid phone number.
                // ValidationPhoneNumber applies international formatting.
                let phoneNumberValidation = ValidationNil() || (ValidationTrim() >>> ValidationPhoneNumber(format: .International))
                phoneNumber = try validate(
                    property: "phoneNumber",
                    with: phoneNumber >>> phoneNumberValidation)
            }
            .append {
                // An email or a phone number is required.
                try validate(
                    properties: ["email", "phoneNumber"],
                    message: "Please provide an email or a phone number.",
                    with: email >>> ValidationNotNil() || phoneNumber >>> ValidationNotNil())
            }
            .validate()
    }
}

var person = Person(name: " Arthur ", age: 35, email: nil, phoneNumber: "0123456789  ")
try person.validate()   // OK
person.name             // "Arthur" (trimmed)
person.phoneNumber      // "+33 1 23 45 67 89" (trimmed & formatted)

var person = Person(name: nil, age: nil, email: "foo@bar.com", phoneNumber: nil)
try person.validate()
// Invalid Person: name should not be empty.

var person = Person(name: "Arthur", age: -1, email: "foo@bar.com", phoneNumber: nil)
try person.validate()
// Invalid Person: age should be greater than or equal to 0.

var person = Person(name: "Arthur", age: 35, email: nil, phoneNumber: nil)
try person.validate()
// Invalid Person: Please provide an email or a phone number.

var person = Person(name: "Arthur", age: 35, email: "foo", phoneNumber: nil)
try person.validate()
// Invalid Person: email is invalid.
```


CHANGELOG.md
Release Notes
=============

## 0.3.0

Released March 1, 2017

**New*

- Conversion to Swift 3 syntax

## 0.2.0

Released August 25, 2015

**Fixed**

- Compiles on Xcode7b6
- Operators `>>>`, `&&` and `||` handle more cases

**New*

- Operator `!` negates a validation
- Operator `~=` provides boolean checks and pattern matching of validations


## v0.1.0

Released on August 23, 2015


ValidationTests/ValidableTests.swift
//
// GRValidation
// https://github.com/groue/GRValidation
// Copyright (c) 2015 Gwendal Roué
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.


import XCTest
import Validation

// A validation that may transform its input.
struct ValidationPhoneNumber : ValidationType {
    enum Format {
        case International
    }
    init(format: Format) { }
    func validate(string: String?) throws -> String {
        let string = try validateNotNil(string)
        return "+33 \(string)"
    }
}

struct SimpleModel : Validable {
    let name: String?
    
    func validate() throws {
        // OK: error is named, and owned.
        try validate(property: "name", with: name >>> ValidationNotNil())
    }
}

struct IntermediateModel : Validable, CustomStringConvertible {
    let name: String?
    let age: Int?
    var phoneNumber: String?
    
    // Ease error description tests by returning a constant description.
    // If this type was a class instead of a struct, errors would not have to
    // store copies, and this problem would be avoided.
    var description: String {
        return "IntermediateModel"
    }
    
    mutating func validate() throws {
        do {
            // OK: readable enough
            // OK: phoneNumber is updated
            // OK: all errors are gathered in a single error
            // FIXME?: ValidationPlan does not adopt ValidationType. This is because we need to mutate self.phoneNumber, and ValidationType is not allowed to perform side effects on value types.
            try ValidationPlan()
                .append { try validate(property: "name", with: name >>> ValidationStringLength(minimum: 1)) }
                .append { try validate(property: "age", with: age >>> ValidationRange(minimum: 0)) }
                .append { phoneNumber = try validate(property: "phoneNumber", with: phoneNumber >>> ValidationPhoneNumber(format: .International)) }
                .validate()
        }
    }
}

struct ComplexModel : Validable {
    let name: String?
    let age: Int?
    let magicWord: String?
    let cardNumber: String?
    let value1: Int?
    let value2: Int?
    
    func validate() throws {
        try ValidationPlan()
            .append { try validate(property: "name", with: name >>> ValidationStringLength(minimum: 1)) }
            .append { try validate(property: "age", with: age >>> ValidationRange(minimum: 0)) }
            .append { try validate(property: "magicWord", with: magicWord >>> (ValidationRegularExpression(pattern: "foo") && ValidationRegularExpression(pattern: "bar"))) }
            .append { try validate(property: "cardNumber", with: cardNumber >>> (ValidationNil<String>() || ValidationStringLength(minimum: 10))) }
            .append { try validate(message: "Value1 or Value2 must be not nil.", with: (value1 >>> ValidationNotNil() || value2 >>> ValidationNotNil())) }
            .validate()
    }
}

struct Person : Validable {
    var name: String?
    var age: Int?
    var email: String?
    var phoneNumber: String?
    
    mutating func validate() throws {
        // ValidationPlan doesn't fail on the first validation error. Instead,
        // it gathers all of them, and eventually throws a single ValidationError.
        try ValidationPlan()
            .append {
                // Name should not be empty after whitespace trimming:
                let nameValidation = ValidationTrim() >>> ValidationStringLength(minimum: 1)
                name = try validate(
                    property: "name",
                    with: name >>> nameValidation)
            }
            .append {
                // Age should be nil, or positive:
                let ageValidation = ValidationNil() || ValidationRange(minimum: 0)
                try validate(
                    property: "age",
                    with: age >>> ageValidation)
            }
            .append {
                // Email should be nil, or contain @ after whitespace trimming:
                let emailValidation = ValidationNil() || (ValidationTrim() >>> ValidationRegularExpression(pattern:"@"))
                email = try validate(
                    property: "email",
                    with: email >>> emailValidation)
            }
            .append {
                // Phone number should be nil, or be a valid phone number.
                // ValidationPhoneNumber applies international formatting.
                let phoneNumberValidation = ValidationNil() || (ValidationTrim() >>> ValidationPhoneNumber(format: .International))
                phoneNumber = try validate(
                    property: "phoneNumber",
                    with: phoneNumber >>> phoneNumberValidation)
            }
            .append {
                // An email or a phone number is required.
                
//                // FIXME: this syntax does not work. Too complex, and nobody can remember it.
//                // TODO: rewrite this in a more straightforward way: if email == nil && phoneNumber == nil { throw ValidationError(...) }
//                try validate(
//                    properties: ["email", "phoneNumber"],
//                    message: "Please provide an email or a phone number.",
//                    with: email >>> ValidationNotNil() || phoneNumber >>> ValidationNotNil())

                if email == nil && phoneNumber == nil {
                    // TODO: This syntax is similar to value validation errors: ValidationError(value: 1, message: "should be nil").
                    // Thus the message should be "requires an email or a phone number", not a full sentence as "Please provide an email or a phone number."
                    // But now, the generated error should then be "Person requires an email or a phone number".
                    // Well, the description of property errors are like "Invalid Person: name should be nil.".
                    // All those sentences are not consistent. We need to fix that.
                    //
                    // TODO: test that this error is associated with both email and phoneNumber properties.
                    throw ValidationError(value: self, message: "Please provide an email or a phone number.", propertyNames: ["email", "phoneNumber"])
                }
                
            }
            .validate()
    }
}

class ValidableTests: ValidationTestCase {
    
    func testSimpleModel() {
        assertNoError {
            let model = SimpleModel(name: "Arthur")
            try model.validate()
        }
        assertNoError {
            let model = SimpleModel(name: nil)
            do {
                try model.validate()
            } catch let error as ValidationError {
                let modelDescription = String(reflecting: model)
                XCTAssertEqual(error.description, "Invalid \(modelDescription): name should not be nil.")
                let nameMessages = error.propertyErrors("name").map { $0.description }.sort()
                XCTAssertEqual(nameMessages, ["Invalid \(modelDescription): name should not be nil."])
            }
        }
    }
    
    func testIntermediateModel() {
        assertNoError {
            var model = IntermediateModel(name:"Arthur", age: 12, phoneNumber: "1 23 45 67 89")
            try model.validate()
            XCTAssertEqual(model.phoneNumber!, "+33 1 23 45 67 89")
        }
        assertNoError {
            var model = IntermediateModel(name:nil, age: 12, phoneNumber: "1 23 45 67 89")
            do {
                try model.validate()
            } catch let error as ValidationError {
                let modelDescription = String(reflecting: model)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(modelDescription): name should not be empty.")
                
                // Test property errors
                let nameMessages = error.propertyErrors("name").map { $0.description }.sort()
                XCTAssertEqual(nameMessages, ["Invalid \(modelDescription): name should not be empty."])
                
                // Test repaired properties
                XCTAssertEqual(model.phoneNumber!, "+33 1 23 45 67 89")
            }
        }
        assertNoError {
            var model = IntermediateModel(name:"Arthur", age: nil, phoneNumber: "1 23 45 67 89")
            do {
                try model.validate()
            } catch let error as ValidationError {
                let modelDescription = String(reflecting: model)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(modelDescription): age should be greater than or equal to 0.")
                
                // Test property errors
                let ageMessages = error.propertyErrors("age").map { $0.description }.sort()
                XCTAssertEqual(ageMessages, ["Invalid \(modelDescription): age should be greater than or equal to 0."])
                
                // Test repaired properties
                XCTAssertEqual(model.phoneNumber!, "+33 1 23 45 67 89")
            }
        }
        assertNoError {
            var model = IntermediateModel(name:nil, age: nil, phoneNumber: nil)
            do {
                try model.validate()
            } catch let error as ValidationError {
                let modelDescription = String(reflecting: model)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(modelDescription): name should not be empty. Invalid \(modelDescription): age should be greater than or equal to 0. Invalid \(modelDescription): phoneNumber should not be nil.")
                
                // Test property errors
                let nameMessages = error.propertyErrors("name").map { $0.description }.sort()
                XCTAssertEqual(nameMessages, ["Invalid \(modelDescription): name should not be empty."])
                let ageMessages = error.propertyErrors("age").map { $0.description }.sort()
                XCTAssertEqual(ageMessages, ["Invalid \(modelDescription): age should be greater than or equal to 0."])
                let phoneNumberMessages = error.propertyErrors("phoneNumber").map { $0.description }.sort()
                XCTAssertEqual(phoneNumberMessages, ["Invalid \(modelDescription): phoneNumber should not be nil."])
            }
        }
    }
    
    func testComplexModel() {
        assertNoError {
            let model = ComplexModel(name: "Arthur", age: 12, magicWord: "foobar", cardNumber: nil, value1: 1, value2: nil)
            try model.validate()
        }
        assertNoError {
            let model = ComplexModel(name: "Arthur", age: 12, magicWord: "fooquxbar", cardNumber: "1234567890", value1: nil, value2: 2)
            try model.validate()
        }
        assertNoError {
            let model = ComplexModel(name: "", age: -12, magicWord: "qux", cardNumber: "123", value1: nil, value2: nil)
            do {
                try model.validate()
            } catch let error as ValidationError {
                let modelDescription = String(reflecting: model)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(modelDescription): name should not be empty. Invalid \(modelDescription): age should be greater than or equal to 0. Invalid \(modelDescription): magicWord is invalid. Invalid \(modelDescription): cardNumber should contain at least 10 characters. Invalid \(modelDescription): Value1 or Value2 must be not nil.")
                
                // Test property errors
                let nameMessages = error.propertyErrors("name").map { $0.description }.sort()
                XCTAssertEqual(nameMessages, ["Invalid \(modelDescription): name should not be empty."])
                let ageMessages = error.propertyErrors("age").map { $0.description }.sort()
                XCTAssertEqual(ageMessages, ["Invalid \(modelDescription): age should be greater than or equal to 0."])
                let magicWordMessages = error.propertyErrors("magicWord").map { $0.description }.sort()
                XCTAssertEqual(magicWordMessages, ["Invalid \(modelDescription): magicWord is invalid."])
                let cardNumberMessages = error.propertyErrors("cardNumber").map { $0.description }.sort()
                XCTAssertEqual(cardNumberMessages, ["Invalid \(modelDescription): cardNumber should contain at least 10 characters."])
                
                // Test global errors
                let modelMessages = error.modelErrors().map { $0.description }.sort()
                XCTAssertEqual(modelMessages, ["Invalid \(modelDescription): Value1 or Value2 must be not nil."])
            }
        }
    }
    
    func testPerson() {
        assertNoError() {
            var person = Person(name: " Arthur ", age: 35, email: nil, phoneNumber: " 1 23 45 67 89 ")
            try person.validate()
            XCTAssertEqual(person.name!, "Arthur")
            XCTAssertEqual(person.phoneNumber!, "+33 1 23 45 67 89")
        }
        assertNoError {
            var person = Person(name: nil, age: nil, email: "foo@bar.com", phoneNumber: nil)
            do {
                try person.validate()
            } catch let error as ValidationError {
                let personDescription = String(reflecting: person)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(personDescription): name should not be empty.")
                
                // Test property errors
                let nameMessages = error.propertyErrors("name").map { $0.description }.sort()
                XCTAssertEqual(nameMessages, ["Invalid \(personDescription): name should not be empty."])
            }
        }
        assertNoError {
            var person = Person(name: "Arthur", age: -1, email: "foo@bar.com", phoneNumber: nil)
            do {
                try person.validate()
            } catch let error as ValidationError {
                let personDescription = String(reflecting: person)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(personDescription): age should be greater than or equal to 0.")
                
                // Test property errors
                let ageMessages = error.propertyErrors("age").map { $0.description }.sort()
                XCTAssertEqual(ageMessages, ["Invalid \(personDescription): age should be greater than or equal to 0."])
            }
        }
        assertNoError {
            var person = Person(name: "Arthur", age: 35, email: "foo", phoneNumber: nil)
            do {
                try person.validate()
            } catch let error as ValidationError {
                let personDescription = String(reflecting: person)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(personDescription): email is invalid.")
                
                // Test property errors
                let emailMessages = error.propertyErrors("email").map { $0.description }.sort()
                XCTAssertEqual(emailMessages, ["Invalid \(personDescription): email is invalid."])
            }
        }
        assertNoError {
            var person = Person(name: "Arthur", age: 35, email: nil, phoneNumber: nil)
            do {
                try person.validate()
            } catch let error as ValidationError {
                let personDescription = String(reflecting: person)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(personDescription): Please provide an email or a phone number.")
                
                // Test property errors
                let emailMessages = error.propertyErrors("email").map { $0.description }.sort()
                XCTAssertEqual(emailMessages, ["Invalid \(personDescription): Please provide an email or a phone number."])
                let phoneNumberMessages = error.propertyErrors("phoneNumber").map { $0.description }.sort()
                XCTAssertEqual(phoneNumberMessages, ["Invalid \(personDescription): Please provide an email or a phone number."])
                
                // Test global errors
                let modelMessages = error.modelErrors().map { $0.description }.sort()
                XCTAssertEqual(modelMessages, ["Invalid \(personDescription): Please provide an email or a phone number."])
            }
        }
        assertNoError {
            var person = Person(name: nil, age: nil, email: nil, phoneNumber: nil)
            do {
                try person.validate()
            } catch let error as ValidationError {
                let personDescription = String(reflecting: person)
                
                // Test full error description
                XCTAssertEqual(error.description, "Invalid \(personDescription): name should not be empty. Invalid \(personDescription): Please provide an email or a phone number.")
                
                // Test property errors
                let nameMessages = error.propertyErrors("name").map { $0.description }.sort()
                XCTAssertEqual(nameMessages, ["Invalid \(personDescription): name should not be empty."])
                let emailMessages = error.propertyErrors("email").map { $0.description }.sort()
                XCTAssertEqual(emailMessages, ["Invalid \(personDescription): Please provide an email or a phone number."])
                let phoneNumberMessages = error.propertyErrors("phoneNumber").map { $0.description }.sort()
                XCTAssertEqual(phoneNumberMessages, ["Invalid \(personDescription): Please provide an email or a phone number."])
                
                // Test global errors
                let modelMessages = error.modelErrors().map { $0.description }.sort()
                XCTAssertEqual(modelMessages, ["Invalid \(personDescription): Please provide an email or a phone number."])
            }
        }
    }
}
